### ISSUE TITLE
query_range API start parameter

### ISSUE THREAD

**Issue by alexandv** [VictoriaMetrics/VictoriaMetrics#669]
**Describe the bug**
When using the query_range api the first returned timestamp does not always match the requested start timestamp. This causes an issue with carbonapi that returns no data.

**To Reproduce**
`curl http://localhost:8428/api/v1/query_range\?query\=%7B__name__%3D\~%22collectd%5C%5C.cpu%5C%5C.19%5C%5C.cpu%5C%5C.idle%22%7D\&start\=1596513601\&step\=60s\&end\=1596513700`
will return a result with the first timestamp set to **1596513601**

However this:
`curl http://localhost:8428/api/v1/query_range\?query\=%7B__name__%3D\~%22collectd%5C%5C.cpu%5C%5C.19%5C%5C.cpu%5C%5C.idle%22%7D\&start\=1596513601\&step\=60s\&end\=1596600000`
will return a result with the first timestamp set to **1596513600**

**Expected behavior**
Not entirely sure what is the correct behavior but the alignment is not consistent and depends on the time range (i.e. if I use the same start date with a different end date then the first value can have a different timestamp).

**Version**
```
/ # ./victoria-metrics-prod --version
victoria-metrics-20200730-173404-tags-v1.39.2-0-ge7959094f
```

**Used command-line flags**
```
"bigMergeConcurrency" = "0"
"csvTrimTimestamp" = "1ms"
"dedup.minScrapeInterval" = "0s"
"deleteAuthKey" = "secret"
"denyQueriesOutsideRetention" = "false"
"enableTCP6" = "false"
"envflag.enable" = "false"
"envflag.prefix" = ""
"fs.disableMmap" = "false"
"graphiteListenAddr" = ":2003"
"graphiteTrimTimestamp" = "1s"
"http.disableResponseCompression" = "false"
"http.maxGracefulShutdownDuration" = "7s"
"http.pathPrefix" = ""
"http.shutdownDelay" = "0s"
"httpAuth.password" = "secret"
"httpAuth.username" = ""
"httpListenAddr" = ":8428"
"import.maxLineLen" = "104857600"
"influxListenAddr" = ""
"influxMeasurementFieldSeparator" = "_"
"influxSkipMeasurement" = "false"
"influxSkipSingleField" = "false"
"influxTrimTimestamp" = "1ms"
"insert.maxQueueDuration" = "1m0s"
"loggerErrorsPerSecondLimit" = "10"
"loggerFormat" = "default"
"loggerLevel" = "INFO"
"loggerOutput" = "stderr"
"maxConcurrentInserts" = "64"
"maxInsertRequestSize" = "33554432"
"maxLabelsPerTimeseries" = "30"
"memory.allowedPercent" = "60"
"metricsAuthKey" = "secret"
"opentsdbHTTPListenAddr" = ""
"opentsdbListenAddr" = ""
"opentsdbTrimTimestamp" = "1s"
"opentsdbhttp.maxInsertRequestSize" = "33554432"
"opentsdbhttpTrimTimestamp" = "1ms"
"pprofAuthKey" = "secret"
"precisionBits" = "64"
"promscrape.config" = ""
"promscrape.config.dryRun" = "false"
"promscrape.config.strictParse" = "false"
"promscrape.configCheckInterval" = "0s"
"promscrape.consulSDCheckInterval" = "30s"
"promscrape.disableCompression" = "false"
"promscrape.disableKeepAlive" = "false"
"promscrape.discovery.concurrency" = "100"
"promscrape.discovery.concurrentWaitTime" = "1m0s"
"promscrape.dnsSDCheckInterval" = "30s"
"promscrape.ec2SDCheckInterval" = "1m0s"
"promscrape.fileSDCheckInterval" = "30s"
"promscrape.gceSDCheckInterval" = "1m0s"
"promscrape.kubernetesSDCheckInterval" = "30s"
"promscrape.maxScrapeSize" = "16777216"
"promscrape.suppressScrapeErrors" = "false"
"relabelConfig" = ""
"retentionPeriod" = "12"
"search.cacheTimestampOffset" = "5m0s"
"search.disableCache" = "false"
"search.latencyOffset" = "30s"
"search.logSlowQueryDuration" = "5s"
"search.maxConcurrentRequests" = "16"
"search.maxExportDuration" = "720h0m0s"
"search.maxLookback" = "0s"
"search.maxPointsPerTimeseries" = "100000"
"search.maxQueryDuration" = "30s"
"search.maxQueryLen" = "100000"
"search.maxQueueDuration" = "10s"
"search.maxStalenessInterval" = "0s"
"search.maxTagKeys" = "secret"
"search.maxTagValues" = "100000"
"search.maxUniqueTimeseries" = "5000000"
"search.minStalenessInterval" = "0s"
"search.resetCacheAuthKey" = "secret"
"selfScrapeInstance" = "self"
"selfScrapeInterval" = "10s"
"selfScrapeJob" = "victoria-metrics"
"smallMergeConcurrency" = "0"
"snapshotAuthKey" = "secret"
"storageDataPath" = "/storage"
"tls" = "false"
"tlsCertFile" = ""
"tlsKeyFile" = "secret"
"version" = "false"
```

**Additional context**
I'm ingesting using the graphite API endpoint and querying using carbonapi. The misalignment between start and the first timestamp causes carbonapi to return no data. I'm not entirely sure if this is a bug in VictoriaMetrics or carbonapi. I also have an issue open in the carbonapi repo.

**Comment by valyala** [comment]
VictoriaMetrics aligns the returned timestamps to multiple of `step` value if the number of returned points per time series exceeds 50. This allows caching query results for the previous time ranges and combining the cached results with results for new data. Without the alignment it would be impossible to combine the cached results with new results. The alignment (and caching) may be disabled either by passing `nocache=1` query arg to `/api/v1/query_range` or by passing `-search.disableCache` command-line arg to VictoriaMetrics during startup.

Let's keep this issue open - probably better solution will be proposed, which allows enabling response caching while disabling timestamps' alignment.

**Comment by valyala** [comment]
BTW, the issue may be solved on carbonapi side by aligning `start` and `end` args to multiples of `step` value. Then the cache may be left enabled on VictoriaMetrics side.

**Comment by alexandv** [comment]
Thanks @valyala . That's consistent with what I'm seeing. It works until `now-49m` but not with `now-50m` (step is 60s).
For now I was able to fix it in carbonapi by updating the start time with the first timestamp received from VictoriaMetrics. 

Aligning start and end to a multiple of step value might be better solution though to take advantage of the cache even for results with less than 50 values.

For reference here is the issue I opened on carbonapi repo: https://github.com/go-graphite/carbonapi/issues/504
